File: Go/XEvent/generic.go
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

package XEvent

import (
	"reflect"

	"github.com/frameworkease/Universal.X.Utility/Go/XLog"
)

func register(manager *Manager, id int, proxy Callback, origin any, once bool, source string) bool {
	manager.Mutex.Lock()
	defer manager.Mutex.Unlock()

	handlers := manager.Handlers[id]

	if manager.Singleton && len(handlers) > 0 {
		XLog.Error("XEvent.%s: singleton mode doesn't allow multiple registrations, id=%v", source, id)
		return false
	}

	cptr := uintptr(reflect.ValueOf(origin).Pointer())
	for i := 0; i < len(handlers); i++ {
		if handlers[i].Origin == cptr {
			return false
		}
	}

	handler := new(Handler)
	handler.Func = proxy
	handler.Origin = cptr
	handler.Once = once
	handlers = append(handlers, handler)
	manager.Handlers[id] = handlers

	return true
}

func unregister(manager *Manager, id int, callback any, source string) bool {
	if callback == nil {
		XLog.Error("XEvent.%s: nil callback, id=%v", source, id)
		return false
	}

	cptr := uintptr(reflect.ValueOf(callback).Pointer())
	manager.Mutex.Lock()
	defer manager.Mutex.Unlock()

	handlers, ok := manager.Handlers[id]
	if !ok {
		return false
	}

	nhandlers := make([]*Handler, 0, len(handlers))
	for i := range handlers {
		if handlers[i].Origin != cptr {
			nhandlers = append(nhandlers, handlers[i])
		}
	}

	sig := len(nhandlers) != len(handlers)
	if sig {
		if len(nhandlers) == 0 {
			delete(manager.Handlers, id)
		} else {
			manager.Handlers[id] = nhandlers
		}
	}

	return sig
}

func RegisterT0(manager *Manager, id int, callback func(), once ...bool) bool {
	if callback == nil {
		XLog.Error("XEvent.RegisterT0: nil callback, id=%v", id)
		return false
	}
	proxy := func(args ...any) { callback() }
	return register(manager, id, proxy, callback, len(once) > 0 && once[0], "RegisterT0")
}

func RegisterT1[T1 any](manager *Manager, id int, callback func(T1), once ...bool) bool {
	if callback == nil {
		XLog.Error("XEvent.RegisterT1: nil callback, id=%v", id)
		return false
	}

	proxy := func(args ...any) {
		var arg1 T1
		if len(args) > 0 && args[0] != nil {
			if v, ok := args[0].(T1); ok {
				arg1 = v
			}
		}
		callback(arg1)
	}

	return register(manager, id, proxy, callback, len(once) > 0 && once[0], "RegisterT1")
}

func RegisterT2[T1, T2 any](manager *Manager, id int, callback func(T1, T2), once ...bool) bool {
	if callback == nil {
		XLog.Error("XEvent.RegisterT2: nil callback, id=%v", id)
		return false
	}

	proxy := func(args ...any) {
		var arg1 T1
		var arg2 T2
		if len(args) > 0 && args[0] != nil {
			if v, ok := args[0].(T1); ok {
				arg1 = v
			}
		}
		if len(args) > 1 && args[1] != nil {
			if v, ok := args[1].(T2); ok {
				arg2 = v
			}
		}
		callback(arg1, arg2)
	}

	return register(manager, id, proxy, callback, len(once) > 0 && once[0], "RegisterT2")
}

func RegisterT3[T1, T2, T3 any](manager *Manager, id int, callback func(T1, T2, T3), once ...bool) bool {
	if callback == nil {
		XLog.Error("XEvent.RegisterT3: nil callback, id=%v", id)
		return false
	}

	proxy := func(args ...any) {
		var arg1 T1
		var arg2 T2
		var arg3 T3
		if len(args) > 0 && args[0] != nil {
			if v, ok := args[0].(T1); ok {
				arg1 = v
			}
		}
		if len(args) > 1 && args[1] != nil {
			if v, ok := args[1].(T2); ok {
				arg2 = v
			}
		}
		if len(args) > 2 && args[2] != nil {
			if v, ok := args[2].(T3); ok {
				arg3 = v
			}
		}
		callback(arg1, arg2, arg3)
	}

	return register(manager, id, proxy, callback, len(once) > 0 && once[0], "RegisterT3")
}

func UnregisterT0(manager *Manager, id int, callback func()) bool {
	return unregister(manager, id, callback, "UnregisterT0")
}

func UnregisterT1[T1 any](manager *Manager, id int, callback func(T1)) bool {
	return unregister(manager, id, callback, "UnregisterT1")
}

func UnregisterT2[T1, T2 any](manager *Manager, id int, callback func(T1, T2)) bool {
	return unregister(manager, id, callback, "UnregisterT2")
}

func UnregisterT3[T1, T2, T3 any](manager *Manager, id int, callback func(T1, T2, T3)) bool {
	return unregister(manager, id, callback, "UnregisterT3")
}

File: CSharp/src/Runtime/Scripts/XEvent/generic.cs
// Copyright (c) 2026 FrameworkEase Technologies. All rights reserved.
// Use of this source code is governed by a MIT-style
// license that can be found in the LICENSE file.

using System;
using System.Collections.Generic;

namespace FrameworkEase.Universal.X.Utility
{
    public partial class XEvent
    {
        public partial class Manager
        {
            internal bool Register(int id, Callback proxy, Delegate origin, bool once, string source)
            {
                if (origin == null)
                {
                    XLog.Error("XEvent.Manager.{0}: nil callback, id={1}.", source, id);
                    return false;
                }
                Mutex.EnterWriteLock();
                try
                {
                    if (Handlers.TryGetValue(id, out List<Handler> handlers) == false)
                    {
                        handlers = new List<Handler>();
                        Handlers.Add(id, handlers);
                    }
                    if (Singleton && handlers.Count > 0)
                    {
                        XLog.Error("XEvent.Manager.{0}: singleton mode doesn't allow multiple registrations, id={1}.", source, id);
                        return false;
                    }
                    for (var i = 0; i < handlers.Count; i++)
                    {
                        if (Equals(handlers[i].Origin, origin)) return false;
                    }
                    var handler = new Handler { Func = proxy, Origin = origin, Once = once };
                    handlers.Add(handler);
                    return true;
                }
                finally { Mutex.ExitWriteLock(); }
            }

            internal bool Unregister(int id, Delegate callback, string source)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.{0}: nil callback, id={1}.", source, id);
                    return false;
                }
                Mutex.EnterWriteLock();
                try
                {
                    var ret = false;
                    if (Handlers.TryGetValue(id, out var handlers))
                    {
                        ret = handlers.RemoveAll(h => Equals(h.Origin, callback)) > 0;
                        if (handlers.Count == 0) Handlers.Remove(id);
                    }
                    return ret;
                }
                finally { Mutex.ExitWriteLock(); }
            }

            public virtual bool Register(Enum id, Action callback, bool once = false) { return Register(id.GetHashCode(), callback, once); }

            public virtual bool Register(int id, Action callback, bool once = false)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.Register: nil callback, id={0}.", id);
                    return false;
                }
                var proxy = new Callback(args => callback?.Invoke());
                return Register(id, proxy, callback, once, "Register");
            }

            public virtual bool Register<T1>(Enum id, Action<T1> callback, bool once = false) { return Register(id.GetHashCode(), callback, once); }

            public virtual bool Register<T1>(int id, Action<T1> callback, bool once = false)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.Register: nil callback, id={0}.", id);
                    return false;
                }
                var proxy = new Callback(args =>
                {
                    var arg1 = args != null && args.Length > 0 ? (T1)args[0] : default;
                    callback?.Invoke(arg1);
                });
                return Register(id, proxy, callback, once, "RegisterT1");
            }

            public virtual bool Register<T1, T2>(Enum id, Action<T1, T2> callback, bool once = false) { return Register(id.GetHashCode(), callback, once); }

            public virtual bool Register<T1, T2>(int id, Action<T1, T2> callback, bool once = false)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.Register: nil callback, id={0}.", id);
                    return false;
                }
                var proxy = new Callback(args =>
                {
                    var arg1 = args != null && args.Length > 0 ? (T1)args[0] : default;
                    var arg2 = args != null && args.Length > 1 ? (T2)args[1] : default;
                    callback?.Invoke(arg1, arg2);
                });
                return Register(id, proxy, callback, once, "RegisterT2");
            }

            public virtual bool Register<T1, T2, T3>(Enum id, Action<T1, T2, T3> callback, bool once = false) { return Register(id.GetHashCode(), callback, once); }

            public virtual bool Register<T1, T2, T3>(int id, Action<T1, T2, T3> callback, bool once = false)
            {
                if (callback == null)
                {
                    XLog.Error("XEvent.Manager.Register: nil callback, id={0}.", id);
                    return false;
                }
                var proxy = new Callback(args =>
                {
                    var arg1 = args != null && args.Length > 0 ? (T1)args[0] : default;
                    var arg2 = args != null && args.Length > 1 ? (T2)args[1] : default;
                    var arg3 = args != null && args.Length > 2 ? (T3)args[2] : default;
                    callback?.Invoke(arg1, arg2, arg3);
                });
                return Register(id, proxy, callback, once, "RegisterT3");
            }

            public virtual bool Unregister(Enum id, Action callback) { return Unregister(id.GetHashCode(), callback); }

            public virtual bool Unregister(int id, Action callback) { return Unregister(id, callback, "Unregister"); }

            public virtual bool Unregister<T1>(Enum id, Action<T1> callback) { return Unregister(id.GetHashCode(), callback); }

            public virtual bool Unregister<T1>(int id, Action<T1> callback) { return Unregister(id, callback, "UnregisterT1"); }

            public virtual bool Unregister<T1, T2>(Enum id, Action<T1, T2> callback) { return Unregister(id.GetHashCode(), callback); }

            public virtual bool Unregister<T1, T2>(int id, Action<T1, T2> callback) { return Unregister(id, callback, "UnregisterT2"); }

            public virtual bool Unregister<T1, T2, T3>(Enum id, Action<T1, T2, T3> callback) { return Unregister(id.GetHashCode(), callback); }

            public virtual bool Unregister<T1, T2, T3>(int id, Action<T1, T2, T3> callback) { return Unregister(id, callback, "UnregisterT3"); }
        }
    }
}
